flows/common/common.go

```diff
@@ -68,36 +80,84 @@ func Handler(fn hdrsFunc, match matchFunc, reporter *Reporter) lwotg.FlowGenerat
 			size := len(buf.Bytes())
 
 			klog.Infof("%s Tx interface %s", flow.Name, tx)
-			handle, err := pcap.OpenLive(tx, 1500, true, pcapTimeout)
+
+			ih, err := pcap.NewInactiveHandle(tx)
+			if err != nil {
+				klog.Errorf("cannot create handle, err: %v", err)
+				return
+			}
+			defer ih.CleanUp()
+			if err := ih.SetImmediateMode(true); err != nil {
+				klog.Errorf("cannot set immediate mode on handle, err: %v", err)
+				return
+			}
+
+			if err := ih.SetPromisc(false); err != nil {
+				klog.Errorf("cannot set promiscous mode, err: %v", err)
+				return
+			}
+
+			if err := ih.SetSnapLen(packetBytes); err != nil {
+				klog.Errorf("cannot set packet length, err: %v", err)
+			}
+
+			// Set buffer to be 200MB of sent packets.
+			if err := ih.SetBufferSize(200e6); err != nil {
+				klog.Errorf("cannot set buffer size, err: %v", err)
+			}
+
+			handle, err := ih.Activate()
 			if err != nil {
 				klog.Errorf("%s Tx error: %v", flow.Name, err)
 				return
 			}
 			defer handle.Close()
 
 			f.setTransmit(true)
+			totPackets := uint32(0)
+			// stopFlow indicates whether we should stop sending packets on the flow, it is set
+			// when the flow specification says that we should only send a limited number of
+			// packets.
+			var stopFlow bool
 			for {
 				select {
 				case <-stop:
 					klog.Infof("%s send exiting on %s", flow.Name, tx)
 					f.setTransmit(false)
 					return
 				default:
-					klog.Infof("%s sending %d packets", flow.Name, pps)
-					for i := 1; i <= int(pps); i++ {
-						if err := handle.WritePacketData(buf.Bytes()); err != nil {
-							klog.Errorf("%s cannot write packet on interface %s, %v", flow.Name, tx, err)
-							return
+					switch stopFlow {
+					case true:
+						// avoid busy looping.
+						time.Sleep(1 * time.Second)
+					default:
+						klog.Infof("%s sending %d packets", flow.Name, pps)
+						sendStart := time.Now()
+						sent := 0
+						for i := 1; i <= int(pps); i++ {
+							if numPackets != 0 && totPackets >= numPackets {
```

Comment:
huh, this naming came out a bit confusing to read. numPackets is the number of packets to send, whereas totPackets is the number of packets that have beens ent so far. Maybe `targetNumPackets` and `sentNumPackets` or something along those lines?

Reply:
ah and we have `sent` as well. maybe sentTotalPackets and sentCurrPackets or something along those lines?

Reply:
Renamed to:

 * `runSentPackets` - and clarified that a "run" is since we started transmitting.
 * `loopSentPackets` for what was done during this loop.

Makes it much clearer, thanks!